corehq/apps/fixtures/upload/workbook.py

```diff
@@ -36,21 +38,31 @@ def __init__(self, file_or_filename):
             self.workbook = excel_get_workbook(file_or_filename)
         except WorkbookJSONError as e:
             raise FixtureUploadError([str(e)])
+        self._rows = {}
         self.item_keys = WeakKeyDictionary()
+        self.ownership = WeakKeyDictionary()
 
     def get_types_sheet(self):
         try:
-            return self.workbook.get_worksheet(title='types')
+            return self.get_data_sheet("types")
         except WorksheetNotFound:
             raise FixtureUploadError([FAILURE_MESSAGES['no_types_sheet']])
 
-    def get_data_sheet(self, data_type_tag):
-        return self.workbook.get_worksheet(data_type_tag)
+    def get_data_sheet(self, tag):
+        # Cache all rows in memory in order to traverse them more than
```

Comment:
I doubt it since rows are usually relatively small objects. The largest lookup tables spreadsheet I have seen is 20MB (from one of the domains for which `SKIP_ORM_FIXTURE_UPLOAD` is enabled). I plan to do some memory profiling with that to verify that this is not a problem.

Reply:
Additionally, without this I could not think of another way to efficiently load owner ids in bulk (rather than on demand as each rows is processed). Currently the old uploader does something like 6 db hits per row while processing ownership. Granted, some of that could be optimized, but it is difficult to make it efficient without multiple passes per sheet.

Reply:
Sounds good.

Reply:
I've done some memory profiling against that 20MB spreadsheet, and while I have not precisely measured the memory impact of caching rows like this, I can confirm that this is by far not the most significant consumer of memory in the uploader. Accumulating `FixtureDataType`, `FixtureDataItem` and `FixtureOwnership` objects in the `CouchTransaction` and then saving them to Couch in bulk takes MUCH more memory (like over 10x).

I have some ideas of how to improve that situation, but I ran into issues with Couch throwing errors, so I may wait to optimize memory usage until after the Couch to SQL migration is complete. In any case caching rows did not cause memory usage to change too much.